### PR TITLE
Remove fully implicit conversions to and from the Fox class

### PR DESCRIPTION
**Pull request by fm3**
We had some trouble with futures not being awaited, fixed in #8513. This was not caught by the compiler because there conveniently was an implicit conversion in scope.

This PR removes fully implicit Fox conversions, to avoid running into this kind of error again.

 - Box, Option and Try can be converted to Fox by calling `toFox` on them
 - Future, Future[Box] and Boolean can be converted to Fox by calling `Fox.fromBool(b)`, `Fox.fromFutureBox(fb)` and `Fox.fromFuture(f)` respectively, forcing users to think which they actually want.
 - Many spots that previously used `futureBox` now use `shiftBox`, which returns `Fox[Box[T]]`, for access to the inner box in fox-style for comprehensions
 - Cleaned up some usages
 - Json parsing is now preferrably done either via PlayBodyParsers or via `JsonHelper.parseAs[T]`, which returns a Box.
 - BoxImplicits are now reduced to the single `bool2Box` (now called `BoxUtils`)
 - Removed some dead code

### URL of deployed dev instance (used for testing):
- https://foxable.webknossos.xyz

### Steps to test:
- This touches almost all our codepaths, but since the compiler is way stricter now I’m fairly confident we didn’t break anything much :tm: 
- Especially test ZipIO (volume upload), loading data from remote datasets, some admin interactions like updating task types, trace a task + view finished as compound
- Just test around a bit

### TODOs:
- [x] find better mechanism for Fox.future2Fox(x.futureBox)
- [x] compiler errors with option2Fox + bool2Fox
- [x] naming Ox/Foxable/…?
- [x] function naming. Fox.fromBool Fox.fromFuture?
- [x] clean up Json.parse usages
- [x] everything else non-implicit
- [x] Loki-Client, code-todos
- [x] move stuff out of implicits that isn’t implicit (object Fox)
- [x] What about the BoxImplicits?
- [x] should Fox.getOrElse return a Fox.successful instead of Future? What about Fox.sequence?
- [x] Does Fox really need the executionContext member?
- [x] Some rudimentary unit tests

------
- [x] Removed dev-only changes like prints and application.conf edits
- [x] Considered [common edge cases](../blob/master/.github/common_edge_cases.md)
- [x] Needs datastore update after deployment

### CODE REVIEW

File: util/src/main/scala/com/scalableminds/util/cache/AlfuCache.scala [hunk 1]
```diff
@@ -1,7 +1,7 @@
 package com.scalableminds.util.cache
 
 import com.github.benmanes.caffeine.cache.{AsyncCache, Caffeine, RemovalCause, RemovalListener, Weigher}
-import com.scalableminds.util.tools.{Fox, FoxImplicits}
+import com.scalableminds.util.tools.{Fox, FoxImplicits, OxImplicits}
```
Comment (philippotto) [inline util/src/main/scala/com/scalableminds/util/cache/AlfuCache.scala:4]: nice :ox:

File: webknossos-tracingstore/app/com/scalableminds/webknossos/tracingstore/tracings/FossilDBClient.scala [hunk 2]
```diff
@@ -78,39 +79,41 @@ class FossilDBClient(collection: String,
...
-              messageWithCauses.append(cause.toString)
-              cause = cause.getCause
-            }
-            new net.liftweb.common.Failure(s"Request to FossilDB failed: $messageWithCauses", Full(e), Empty)
-        }
-        Future.successful(box)
+    Fox.fromFutureBox {
+      future.transformWith {
+        case Success(value) =>
+          Fox.successful(value).futureBox
```
Comment (MichaelBuessemeyer) [inline webknossos-tracingstore/app/com/scalableminds/webknossos/tracingstore/tracings/FossilDBClient.scala:85]: Very nitpicky and this wasn't part of your pr but:

Hmm here you construct a fox to get the future box and then construct a fox again. Below a future box is constructed instead.

So maybe something like `Future.successful(Full(value))` might be more consistent with the `Future.successful(box)` below

**Review by MichaelBuessemeyer — COMMENTED** [review]
Awesome :tada: 

I took a look at all changes and a few nitpicky things & questions.

Note: I did not test yet. Will do that in some near future. (re-requesting review to not forget that)

File: util/src/main/scala/com/scalableminds/util/tools/Fox.scala [hunk 3]
```diff
@@ -1,63 +1,54 @@
...
-      }
-      f <- fut
-    } yield f
-
-  implicit def option2Fox[T](b: Option[T])(implicit ec: ExecutionContext): Fox[T] =
-    new Fox(Future.successful(Box(b)))
-
-  implicit def futureOption2Fox[T](f: Future[Option[T]])(implicit ec: ExecutionContext): Fox[T] =
-    new Fox(f.map(Box(_)))
+  implicit protected def box2Ox[T](b: Box[T]): Foxable[T] =
```
[the remark this answers is not in the document]
Comment (fm3) [reply util/src/main/scala/com/scalableminds/util/tools/Fox.scala:17]: Oops, i renamed only the class to Foxable, but missed the function names. While I do like the Ox, I think the master branch is not the best place for it :innocent:  I’ll rename them to box2Foxable

File: app/controllers/AuthenticationController.scala [hunk 4]
```diff
@@ -586,22 +585,20 @@ class AuthenticationController @Inject()(
                                    email: String): Fox[(String, String, String, List[String])] = {
     var (errors, fN, lN) = normalizeName(firstName, lastName)
     for {
-      nameEmailErrorBox: Box[(String, String, String, List[String])] <- multiUserDAO
-        .findOneByEmail(email.toLowerCase)(GlobalAccessContext)
-        .futureBox
-        .flatMap {
-          case Full(_) =>
-            errors ::= "user.email.alreadyInUse"
+      nameEmailError: (String, String, String,
```
[the remark this answers is not in the document]
Comment (fm3) [reply app/controllers/AuthenticationController.scala:588]: Yup, I think it’s way clearer now. Previously, back-and-forth implicit conversions and type hints produced this weird combination.

File: app/models/annotation/AnnotationInformationProvider.scala [hunk 5]
```diff
@@ -62,15 +62,13 @@ class AnnotationInformationProvider @Inject()(
...
-    val annotationFox = annotationDAO.findOneByTracingId(tracingId)
+  def annotationForTracing(tracingId: String)(implicit ctx: DBAccessContext): Fox[Annotation] =
     for {
-      annotationBox <- annotationFox.futureBox
-    } yield {
-      annotationBox match {
-        case Full(_) => annotationBox
-        case _       => annotationStore.findCachedByTracingId(tracingId)
+      annotationBox <- annotationDAO.findOneByTracingId(tracingId).shiftBox
+      withFallback <- annotationBox match {
```
[the remark this answers is not in the document]
Comment (fm3) [reply app/models/annotation/AnnotationInformationProvider.scala:68]: good point. I now renamed the variables to `storedAnnotationIdBox` and `storedOrCachedAnnotationId`

File: app/models/annotation/AnnotationUploadService.scala [hunk 6]
```diff
@@ -33,20 +33,21 @@ case class SharedParsingParameters(useZipName: Boolean,
...
+      ctx: DBAccessContext): Future[NmlParseResult] =
     extractFromNml(new FileInputStream(file), name, sharedParsingParameters)
 
   private def extractFromNml(inputStream: InputStream,
                              name: String,
                              sharedParsingParameters: SharedParsingParameters,
                              basePath: Option[String] = None)(implicit m: MessagesProvider,
                                                               ec: ExecutionContext,
-                                                              ctx: DBAccessContext): Fox[NmlParseResult] = {
+                                                              ctx: DBAccessContext): Future[NmlParseResult] = {
```
[the remark this answers is not in the document]
Comment (fm3) [reply app/models/annotation/AnnotationUploadService.scala:50]: Exactly right.

Though of course the usage turns it into a fox for compatibility reasons anyway, so it’s not super important.

File: util/src/main/scala/com/scalableminds/util/tools/Fox.scala [hunk 7]
```diff
@@ -334,6 +282,20 @@ class Fox[+A](val futureBox: Future[Box[A]])(implicit ec: ExecutionContext) {
...
+        case Success(resultBox)    => f(resultBox)
+        case scala.util.Failure(e) => f(Failure(e.toString, Some(e), None))
+      }
+    }
+
+  /*
+   * Returns new Fox[Box[A]] that is always successful, such that the original’s box is shifted “inwards”.
+   * Use to access the box.
+   */
+  def shiftBox: Fox[Box[A]] = Fox.fromFuture(futureBox)
```
[the remark this answers is not in the document]
Comment (fm3) [reply util/src/main/scala/com/scalableminds/util/tools/Fox.scala:297]: Hm, I’m not convinced. rebox is nicely short but I think it’s not clearer than shiftBox. shiftBoxInward is more explicit but also very long. I think my current favorite is still shiftBox. Unless you veto it?

File: util/src/main/scala/com/scalableminds/util/tools/Fox.scala [hunk 8]
```diff
@@ -1,63 +1,54 @@
...
+
+  /**
+    * Transform a Future[T] into a Fox[T] such that if the Future contains an exception, it is turned into a Fox.failure
+    */
+  def fromFuture[T](f: Future[T])(implicit ec: ExecutionContext): Fox[T] =
+    fromFutureBox(
+      for {
+        fut <- f.transform {
+          case Success(value)        => Try(Fox.successful(value))
+          case scala.util.Failure(e) => Try(Fox.failure(e.getMessage, Full(e)))
```
[the remark this answers is not in the document]
Comment (fm3) [reply util/src/main/scala/com/scalableminds/util/tools/Fox.scala:47]: `Future.transform` is an overloaded function. The one whose docstring you posted is not the one that is used here.

Instead, we use `def transform[S](f: Try[T] => Try[S])`, which creates a new `Future` from the `Try` returned in the block. So what we’re creating in the block is a `Try[Fox]`, which in turn makes the whole `transform` function return a `Future[Fox]`. On this one we need to map again, so that we hold a `Future[Box]`, which we can then pass to `Fox.fromFutureBox`. All clear? :sweat_smile: 

I’ll rename the local variables to match the types better.

File: util/src/main/scala/com/scalableminds/util/tools/Fox.scala [hunk 9]
```diff
@@ -1,63 +1,54 @@
...
+        }
+        f <- fut.futureBox
```
[the remark this answers is not in the document]
Comment (fm3) [reply util/src/main/scala/com/scalableminds/util/tools/Fox.scala]: See [message above](https://github.com/scalableminds/webknossos/pull/8521#discussion_r2048611401)

**Comment by fm3** [comment]
Thanks for having such a close look! I implemented your suggestions (except for the renaming of shiftBox, see comment). Feel free to test and have a look at my latest commits and answers :)

File: util/src/main/scala/com/scalableminds/util/tools/Fox.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (MichaelBuessemeyer) [reply util/src/main/scala/com/scalableminds/util/tools/Fox.scala]: image

File: util/src/main/scala/com/scalableminds/util/tools/Fox.scala (hunk 7 above)
[the remark this answers is not in the document]
Comment (MichaelBuessemeyer) [reply util/src/main/scala/com/scalableminds/util/tools/Fox.scala:297]: Ok then lets keep `shiftBox`. It also took me some time to get used to `futureBox` so this should be the similar here :D
